mm2src/mm2_net/src/wasm/body_stream.rs

```diff
@@ -0,0 +1,411 @@
+                    }
+                },
+                ReadState::Data(data_length) => {
+                    let data_length = *data_length as usize;
+
+                    if this.buf.len() < data_length {
+                        // Can't read data right now
+                        return Ok(());
+                    } else {
+                        this.incomplete_data.unsplit(this.buf.take(data_length));
+
+                        let new_data = this.incomplete_data.split();
+
+                        if let Some(data) = this.data {
+                            data.unsplit(new_data);
+                        } else {
+                            *this.data = Some(new_data);
+                        }
+
+                        *this.state = ReadState::CompressionFlag;
+                    }
+                },
+                ReadState::TrailerLength => {
+                    if this.buf.len() < 4 {
+                        // Can't read data length right now
+                        return Ok(());
+                    } else {
+                        let trailer_length_bytes = this.buf.take(4);
+                        let trailer_length = BigEndian::read_u32(trailer_length_bytes.as_ref());
+                        *this.state = ReadState::Trailer(trailer_length);
+                    }
+                },
+                ReadState::Trailer(trailer_length) => {
+                    let trailer_length = *trailer_length as usize;
+
+                    if this.buf.len() < trailer_length {
+                        // Can't read trailer right now
+                        return Ok(());
+                    } else {
+                        let mut trailer_bytes = this.buf.take(trailer_length);
+                        trailer_bytes.put_u8(b'\n');
+
+                        let mut trailers_buf = [EMPTY_HEADER; 64];
+                        let parsed_trailers = match httparse::parse_headers(&trailer_bytes, &mut trailers_buf)
+                            .map_err(|err| PostGrpcWebErr::InvalidRequest(err.to_string()))?
+                        {
+                            Status::Complete((_, headers)) => Ok(headers),
+                            Status::Partial => Err(PostGrpcWebErr::InvalidRequest("Invalid".to_string())),
+                        }?;
+
+                        let mut trailers = HeaderMap::with_capacity(parsed_trailers.len());
+
+                        for parsed_trailer in parsed_trailers {
+                            let header_name = HeaderName::from_bytes(parsed_trailer.name.as_bytes())
+                                .map_err(|err| PostGrpcWebErr::InvalidRequest(err.to_string()))?;
+                            let header_value = HeaderValue::from_bytes(parsed_trailer.value)
+                                .map_err(|err| PostGrpcWebErr::InvalidRequest(err.to_string()))?;
+                            trailers.insert(header_name, header_value);
+                        }
```

Comment:
Can this be made into a separate function maybe, called `parse_trailers`?

Reply:
done